src/libstd/sys/unix/net.rs

```diff
@@ -66,7 +67,30 @@ impl Socket {
     }
 
     pub fn duplicate(&self) -> io::Result<Socket> {
-        let fd = try!(cvt(unsafe { libc::dup(self.0.raw()) }));
+        use libc::funcs::posix88::fcntl::fcntl;
+        const ZERO: c_int = 0;
+        #[cfg(target_os = "linux")]
+        fn fcntl_dupfd() -> c_int {
+            const MINUS_ONE: c_int = -1;
+            static START: Once = Once::new();
+            static mut FCNTL_DUPFD: c_int = 0xdead;
+            START.call_once(|| unsafe {
+                let result = fcntl(MINUS_ONE, libc::F_DUPFD_CLOEXEC, ZERO);
+                assert!(result == libc::EBADF || result == libc::EINVAL);
+                if result == libc::EINVAL {
+                    FCNTL_DUPFD = libc::F_DUPFD_CLOEXEC;
+                } else {
+                    FCNTL_DUPFD = libc::F_DUPFD;
+                }
+            });
+            unsafe { FCNTL_DUPFD }
+        }
```

Comment:
Can you add a comment explaining what this function is doing? I'm also somewhat wary about the `assert!` here as there's probably some random kernel which returns some obscure error code. Are we also guaranteed that `EINVAL` is **never** returned for a bad file descriptor?

I'm not sure if there's some prior art to draw from in this sort of detection, but I'm a little wary of this...